drivers/usb/udc/udc_dwc2.c

```diff
@@ -1848,14 +1849,14 @@ static int udc_dwc2_init_controller(const struct device *dev)
 	case USB_DWC2_GHWCFG2_HSPHYTYPE_ULPI:
 		gusbcfg |= USB_DWC2_GUSBCFG_PHYSEL_USB20 |
 			   USB_DWC2_GUSBCFG_ULPI_UTMI_SEL_ULPI;
-		dcfg |= USB_DWC2_DCFG_DEVSPD_USBHS20
-			<< USB_DWC2_DCFG_DEVSPD_POS;
+		dcfg |= usb_dwc2_set_dcfg_devspd(USBD_SUPPORTS_HIGH_SPEED ?
+		    USB_DWC2_DCFG_DEVSPD_USBHS20 : USB_DWC2_DCFG_DEVSPD_USBFS20);
 		break;
 	case USB_DWC2_GHWCFG2_HSPHYTYPE_UTMIPLUS:
 		gusbcfg |= USB_DWC2_GUSBCFG_PHYSEL_USB20 |
 			   USB_DWC2_GUSBCFG_ULPI_UTMI_SEL_UTMI;
-		dcfg |= USB_DWC2_DCFG_DEVSPD_USBHS20
-			<< USB_DWC2_DCFG_DEVSPD_POS;
+		dcfg |= usb_dwc2_set_dcfg_devspd(USBD_SUPPORTS_HIGH_SPEED ?
+		    USB_DWC2_DCFG_DEVSPD_USBHS20 : USB_DWC2_DCFG_DEVSPD_USBFS20);
```

Comment:
This layer violation. You may not use anything from subsys/usb/device_next. And there is absolutely no need to limit the high-speed controller to full speed.

Reply:
There are valid reasons for disabling high-speed, otherwise the peripherals wouldn't support it.

Reply:
Setting the reasons aside (the most trivial one is replicating an existing device with a newer SoC in a way that allows reusing existing host software without any modifications), please guide how to achieve this functionality without "layer violation". Note that USB itself has design choices that could be classified as layering violations, like:
* different configurations based on operating speed
* the meaning of bInterval field in Endpoint descriptor depends on operating speed
* isochronous feedback size and meaning depends on operating speed

The operating speed plays so significant role in USB that I believe there must be a way to force Full-Speed only operation of High-Speed capable device without resorting to filtering out high-speed detection handshake on D+/D- lines.